feat(nms): Add Domain Proxy UI

IgoninVV opened the pull request:
## Summary
Add CBSDs page in Equpmint section
<img width="1792" alt="cbsds" src="https://user-images.githubusercontent.com/10230636/170262891-45ac6e2a-a507-4040-be6c-df1486875c9e.png">


Add create/edit dialog for CBSD
<img width="968" alt="add-sbsd-1" src="https://user-images.githubusercontent.com/10230636/170263007-37dc81b6-d916-41bf-915f-374a60536a1d.png">
<img width="965" alt="add-sbsd-2" src="https://user-images.githubusercontent.com/10230636/170263019-feadfa55-cd1c-4937-a0fc-cef7dd9f4609.png">

Add Domain Proxy Logs page in Metric section
<img width="1792" alt="dp-logs-1" src="https://user-images.githubusercontent.com/10230636/170263051-fc37ea7a-e710-4b0d-84d0-849bb078fcff.png">
<img width="1792" alt="dp-logs-2" src="https://user-images.githubusercontent.com/10230636/170263067-1b8a4e50-384c-4d07-b71d-81eb9b1cd240.png">

maxhbr commented:
since that PR adds a feature to nms, it would probably be more appropriate to call it `feat(nms):...`, right?

maxhbr commented on nms/app/components/context/CbsdContext.js:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Copyright 2020 The Magma Authors.
```
2022?

thmsschmitt replied:
At moment all headers are for 2020 and there is a check with a hard coded year.

Neudrino replied:
Thanks to https://github.com/magma/magma/pull/12714 this is now possible to hold an up to date date. :wink:

ivan-freedomfi replied:
Changed Copyright to 2022 in all files touched :)

thmsschmitt replied:
@maxhbr correct me if I'm wrong, but shouldn't only new files get the 2022 header?

thmsschmitt commented on nms/app/views/domain-proxy/CbsdsTable.js:
```diff
@@ -0,0 +1,205 @@
+function CbsdsTable(props: WithAlert) {
+  const enqueueSnackbar = useEnqueueSnackbar();
+
+  const [openEditDialog, setOpenEditDialog] = useState(false);
```
nit: `isEditDialodOpen` and `setIsEditDialogOpen`

IgoninVV replied:
Fixed

thmsschmitt commented on nms/app/views/domain-proxy/CbsdEdit.js:
```diff
@@ -0,0 +1,425 @@
+  }
+}
+
+export function CbsdAddEditDialog(props: DialogProps) {
```
We should have some place holders here. Ideally with example input
![Screenshot 2022-05-05 at 18 24 56](https://user-images.githubusercontent.com/102796295/166969291-ed1b692b-3256-49dc-b47f-f1e99c395243.png)
.

IgoninVV replied:
Fixed

thmsschmitt replied:
I still see a bunch of blank fields. Checkout the "Add Gateway" Dialog for how this should ideally look like. 
It is especially useful if we require a certain format, but I can't judge if this is the case here.
![Screenshot 2022-06-13 at 09 12 35](https://user-images.githubusercontent.com/102796295/173300006-98585731-b310-4dd5-a44d-563870a492d0.png)

ivan-freedomfi replied:
Indeed I can still see 3 blank fields, and the other ones have wrong color: probably those are not placeholders but default values.

## Is this what you mean?
<img width="1597" alt="image" src="https://user-images.githubusercontent.com/104149278/173365206-846ce29b-3c1d-4584-845e-725bd5ad781a.png">

thmsschmitt replied:
Yes. Although the zero for Number of Antennas doesn't have the wrong color imho. If if is a valid pre-filled value it is fine. If it isn't valid it should be replaced with something like `Enter number of antennas`

ivan-freedomfi replied:
I've added 3 placeholders:

<img width="878" alt="image" src="https://user-images.githubusercontent.com/104149278/173629373-eeaa9afc-f2b1-4ca3-a57d-1cf1e71c4c26.png">

thmsschmitt commented on nms/app/views/domain-proxy/CbsdEdit.js:
```diff
@@ -0,0 +1,425 @@
+    setCbsdFormData(newValue);
+  }, [props.open, props.cbsd]);
+
+  useEffect(() => {
```
Maybe I'm missing something here. What is the purpose of this effect? `formErrors` are local state anyhow.

thmsschmitt replied:
I see the intention was probably have `props.open` in the dependencies to reset the error on close which does not happen at the moment.

IgoninVV replied:
Fixed

thmsschmitt commented on nms/app/views/domain-proxy/CbsdEdit.js:
```diff
@@ -0,0 +1,425 @@
+    try {
+      setIsLoading(true);
+
+      const cbsdData: mutable_cbsd = {
```
`desired_state` is missing see also https://github.com/magma/magma/runs/6319913366?check_suite_focus=true

IgoninVV replied:
Fixed

thmsschmitt commented on nms/app/views/domain-proxy/CbsdEdit.js:
```diff
@@ -0,0 +1,425 @@
+  return (
+    <>
+      <CbsdAddEditDialog open={open} onClose={handleClose} />
+      {props.isLink ? (
```
`isLink` is always false

IgoninVV replied:
Fixed

thmsschmitt commented on nms/app/views/domain-proxy/CbsdEdit.js:
```diff
@@ -0,0 +1,425 @@
+      background: colors.primary.mirage,
+    },
+  },
+  tabBar: {
```
Unused

IgoninVV replied:
Fixed

thmsschmitt commented on nms/app/components/lte/LteContext.js:
```diff
@@ -138,6 +142,97 @@ export function GatewayContextProvider(props: Props) {
+    refetch();
+  }, [
+    refetch,
+    paginationOptions.page,
```
Technically we don't need `paginationOptions` as dependencies because `refetch` will change when the `paginationOptions` change. But I don't mind keeping this to make it more obvious under which circumstances we refetch.

ivan-freedomfi replied:
So we leave this as is? :)
Sorry it's not my code, the author left the company and I'm finishing for him :)

thmsschmitt replied:
Fine for me

thmsschmitt commented on nms/app/views/domain-proxy/CbsdEdit.js:
```diff
@@ -0,0 +1,425 @@
+                              });
+                            }}
+                          >
+                            <DeleteIcon />
```
Would be nice if the delete icon would be vertically centred
![Screenshot 2022-05-10 at 16 01 46](https://user-images.githubusercontent.com/102796295/167654202-afe9b80c-5c2f-4651-b53f-dbc701983217.png)

IgoninVV replied:
Fixed

thmsschmitt commented on nms/app/views/domain-proxy/LogsList.js:
```diff
@@ -0,0 +1,418 @@
+          <CardTitleRow
+            key="title"
+            icon={ListIcon}
+            label={`Doman Proxy Logs`}
```
`Doman` -> `Domain`

IgoninVV replied:
Fixed

thmsschmitt commented on nms/app/views/domain-proxy/LogsList.js:
```diff
@@ -0,0 +1,418 @@
+  );
+}
+
+export default withAlert(LogsList);
```
The `WithAlert` HOC is not used

IgoninVV replied:
Fixed

thmsschmitt commented on nms/app/views/domain-proxy/CbsdEdit.js:
```diff
@@ -0,0 +1,425 @@
+    }
+  };
+
+  return (
```
The NMS does have a very bad test test coverage. I want to change this.
Please add some basic test e.g. for editing and creating a CBSD.

ivan-freedomfi replied:
Tests added, please take a look

thmsschmitt commented on nms/app/views/domain-proxy/LogsList.js:
```diff
@@ -0,0 +1,418 @@
+  const [responseCode, setResponseCode] = useState < string | null > (null);
+  const [logName, setLogName] = useState < string > ('');
+
+  const getDataFn = useCallback((query) => {
```
Also here some test for the filtering would be nice.

ivan-freedomfi replied:
Logs filtering test added, please take a look

thmsschmitt reviewed (changes requested):
Please address the review comments. This also needs a rebase. And the linter is rather unhappy https://github.com/magma/magma/runs/6319913140?check_suite_focus=true

Neudrino reviewed (commented):
Linter and formatter should be run. Often there is the EOL at EOF missing.

IgoninVV committed:
feat(nms): Add Domain Proxy UI (CBSDs and logs)

- Add CRUD for CBSDs under Equipment page.
- Add Domain Proxy logs page under Metrics page.

Signed-off-by: Ivan Sergiienko <ivan@freedomfi.com>

thmsschmitt commented on nms/app/views/domain-proxy/CbsdsTable.js:
```diff
@@ -0,0 +1,216 @@
+        pageSize: query.pageSize,
+      });
+
+      return new Promise(async resolve => {
```
Nit: Can be shortened to `Promise.resolve({ data, ...`

ivan-freedomfi replied:
👍 corrected

thmsschmitt commented on nms/app/views/domain-proxy/LogsList.js:
```diff
@@ -0,0 +1,494 @@
+
+  const getDataFn = useCallback(
+    query => {
+      return new Promise(async (resolve, reject) => {
```
We should not mix explicit promise creation and async functions.
We can just omit the  promise creation here and throw an error instead of calling reject.

ivan-freedomfi replied:
Well noticed! I've simplified it to use `async/await` only

thmsschmitt commented on nms/app/views/domain-proxy/__tests__/CbsdEdit.js:
```diff
@@ -0,0 +1,253 @@
+  });
+
+  afterEach(() => {
+    MagmaAPIBindings.getDpByNetworkIdCbsds.mockClear();
```
We no longer need to do this manually, this now happens automatically.

ivan-freedomfi replied:
Hm indeed it works without this line.
I find it confusing though, if I mock resolved value in `beforeEach`, I always reset it in `afterEach`.
So the code becomes more weird without explicit resets.

Do you really want me to remove this `mockClear()` ? :)

thmsschmitt replied:
Yes we have changed this here https://github.com/magma/magma/pull/12718
This makes the tests more robust because no test can silently rely on a mock that is called or setup by another test.

ivan-freedomfi replied:
Interesting :)
I've removed all `.mockClear()`, it works without it indeed! :)

thmsschmitt commented on nms/app/views/domain-proxy/__tests__/CbsdsTable.js:
```diff
@@ -0,0 +1,141 @@
+
+    const dialog = await component.findByRole('dialog');
+
+    await within(dialog).findByText('Edit CBSD');
```
A test without expect doesn't communicate well what is tested

ivan-freedomfi replied:
Agreed, changed to
```
expect(within(dialog).getByText('Edit CBSD')).toBeInTheDocument();
```

thmsschmitt reviewed (approved):
Looks good to me, I left some minor review comments that should be addressed. 
Also the linter and flow check are red and need to be fixed.

ivan-freedomfi commented on nms/flow-typed/npm/@testing-library/react_v10.x.x.js:
```diff
@@ -1,19 +1,23 @@
-// flow-typed signature: 38ccc5bbcbb54d21f37adca5cc3ed7a1
-// flow-typed version: 0b0eb4afe9/@testing-library/react_v9.x.x/flow_>=v0.104.x
+// flow-typed signature: e9406c592c1d0b6096088adad78582d2
```
I've updated types for `@testing-library/react` because `findByText` and `screen` were not available in the old version. But these functions were available in the library itself, the problem was only with the types.

It seems this upgrade doesn't break anything else.

Are you ok with it?

thmsschmitt replied:
No that is absolutely fine. We are in the process of [migrating the NMS to TypeScript](https://github.com/orgs/magma/projects/14) so this will anyhow soon be obsolete.

ivan-freedomfi replied:
Great! :)
The Github project link is 404 for me, JFYI :)

thmsschmitt replied:
The link should work now

ivan-freedomfi replied:
👍 the link works now, thanks
